gitlab/utils.py

```diff
@@ -56,7 +56,88 @@ def copy_dict(dest: Dict[str, Any], src: Dict[str, Any]) -> None:
             dest[k] = v
 
 
-def _url_encode(id: str) -> str:
+class EncodedId(str):
+    """A custom `str` class that will return the URL-encoded value of the string.
+
+    Features:
+      * Using it recursively will only url-encode the value once.
+      * Can accept either `str` or `int` as input value.
+      * Can be used in an f-string and output the URL-encoded string.
+
+    Reference to documentation on why this is necessary.
+
+    https://docs.gitlab.com/ee/api/index.html#namespaced-path-encoding
+
+    If using namespaced API requests, make sure that the NAMESPACE/PROJECT_PATH is
+    URL-encoded. For example, / is represented by %2F
+
+    https://docs.gitlab.com/ee/api/index.html#path-parameters
+
+    Path parameters that are required to be URL-encoded must be followed. If not, it
+    doesn’t match an API endpoint and responds with a 404. If there’s something in
+    front of the API (for example, Apache), ensure that it doesn’t decode the
+    URL-encoded path parameters.
+
+
+    When creating an EncodedId instance `__new__` will be called first and then
+    `__init__`.
+    """
```

Comment:
Any chance we can get more of this on the screen/intellisense, e.g. by removing upstream text that might get outdated? And are we not documenting the default behavior of `__new__` and `__init__` here?

```suggestion
    """A custom `str` class that will return the URL-encoded value of the string:

      * Using it recursively will only url-encode the value once.
      * Can accept either `str` or `int` as input value.
      * Can be used in an f-string and output the URL-encoded string.

    See:

      * https://docs.gitlab.com/ee/api/index.html#namespaced-path-encoding
      * https://docs.gitlab.com/ee/api/index.html#path-parameters
    """
```

![image](https://user-images.githubusercontent.com/16777978/148859757-16a02c40-9efd-4105-815f-efb457ff7bda.png)

Reply:
Sure I can work on that. Thanks.

Reply:
> Any chance we can get more of this on the screen/intellisense, e.g. by removing upstream text that might get outdated? 

Sure.

> And are we not documenting the default behavior of `__new__` and `__init__` here?

You're a better Pythonista than me if you already knew this 😊  But I guess we can let me people who don't know that go figure it out.

Reply:
Done.